zerver/lib/upload.py

```diff
@@ -385,14 +387,31 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
     return uploaded_file_name, uploaded_file_size, content_type
 
 
+@contextlib.contextmanager
+def maybe_unset_proxy():
+    # https://github.com/boto/botocore/issues/2644 means that the IMDS
+    # request _always_ pulls from the environment.  Monkey-patch the
+    # `should_bypass_proxies` function if we need to skip them, based
+    # on S3_SKIP_PROXY.
+    with contextlib.ExitStack() as stack:
+        if settings.S3_SKIP_PROXY == True or (
+            settings.S3_SKIP_PROXY is None and not settings.S3_KEY
+        ):
+            stack.enter_context(
+                mock.patch("botocore.utils.should_bypass_proxies", side_effect=lambda url: True)
```

Comment:
`mock.patch` isn’t thread-safe either. Not sure there’s any advantage in this over just doing `botocore.utils.should_bypass_proxies = lambda url: True` at top level.